core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -416,6 +416,9 @@ public void updateRoutingEntry(ShardRouting newRouting) throws IOException {
                 // active primaries.
                 throw new IndexShardRelocatedException(shardId(), "Shard is marked as relocated, cannot safely move to state " + newRouting.state());
             }
+            assert newRouting.active() == false || state == IndexShardState.STARTED || state == IndexShardState.RELOCATED ||
+                state == IndexShardState.CLOSED :
+                "shard state is " + state + ", but routing is active " + newRouting;
```

Comment:
can you elaborate? What's not correct?
What I want to check is `newRouting.active() ==> state == XYZ`, which is rewritten as `newRouting.active() == false || state == XYZ`